lms/djangoapps/courseware/masquerade.py

```diff
@@ -40,6 +40,16 @@ def __init__(self, course_key, role='student', user_partition_id=None, group_id=
         self.group_id = group_id
         self.user_name = user_name
 
+    def __setstate__(self, state):
+        """
+        Ensure that all attributes are initialised when unpickling CourseMasquerade objects.
+
+        Users might still have CourseMasquerade objects from older versions of the code in their
+        session.  These old objects might not have all attributes set, possibly resulting in
+        AttributeErrors.
+        """
+        self.__init__(**state)
```

Comment:
Can __setstate__ be called with any parameters outside of the __init__ defined keyword args? I'm not familiar with __setstate__() but I'm wondering if an unexpected parameter gets passed in here, __init__ with that unexpected parameter will be called which will throw an exception?